rust/lance-index/src/scalar/btree.rs

```diff
@@ -0,0 +1,809 @@
+        }
+    }
+
+    // All pages that could have a value in the range
+    fn pages_between(&self, range: (Bound<ScalarValue>, Bound<ScalarValue>)) -> Vec<u64> {
+        self.tree
+            .range((Self::wrap_bound(range.0), Self::wrap_bound(range.1)))
+            .map(|(_, val)| val)
+            .flatten()
+            .copied()
+            .collect::<Vec<_>>()
+    }
+
+    fn pages_null(&self) -> Vec<u64> {
+        self.null_pages.clone()
+    }
+}
+
+/// A trait that must be implemented by a btree sub-index
+///
+/// A sub-index must be capable of indexing a single page of data.  We represent
+/// pages as a single record batch.
+#[async_trait]
+pub trait SubIndexLoader: std::fmt::Debug + Send + Sync {
+    async fn load_subindex(
+        &self,
+        offset: u64,
+        index_reader: Arc<dyn IndexReader>,
+    ) -> Result<Arc<dyn ScalarIndex>>;
+}
+
+/// A btree index satisfies scalar queries using a b tree
+///
+/// The upper layers of the btree are expected to be cached and, when unloaded,
+/// are stored in a btree structure in memory.  The leaves of the btree are left
+/// to be searched by some other kind of index (currently a flat search).
+///
+/// This strikes a balance between an expensive memory structure containing all
+/// of the values and an expensive disk structure that can't be efficiently searched.
+///
+/// For example, given 1Bi values we can store 256Ki leaves of size 4Ki.  We only
+/// need memory space for 256Ki leaves (depends on the data type but usually a few MiB
+/// at most) and can narrow our search to 4Ki values.
+///
+/// Note: this is very similar to the IVF index except we store the IVF part in a btree
+/// for faster lookup
+#[derive(Debug)]
+pub struct BTreeIndex {
+    page_lookup: BTreeLookup,
+    store: Arc<dyn IndexStore>,
+    sub_index: Arc<dyn SubIndexLoader>,
+}
+
+impl BTreeIndex {
+    fn new(
+        tree: BTreeMap<OrderableScalarValue, Vec<u64>>,
+        null_pages: Vec<u64>,
+        store: Arc<dyn IndexStore>,
+        sub_index: Arc<dyn SubIndexLoader>,
+    ) -> Self {
+        let page_lookup = BTreeLookup::new(tree, null_pages);
+        Self {
+            page_lookup,
+            store,
+            sub_index,
+        }
+    }
+
+    async fn search_page(
+        &self,
+        query: &ScalarQuery,
+        page_offset: u64,
+        index_reader: Arc<dyn IndexReader>,
+    ) -> Result<UInt64Array> {
+        let subindex = self
+            .sub_index
+            .load_subindex(page_offset, index_reader)
+            .await?;
+        // TODO: If this is an IN query we can perhaps simplify the subindex query by restricting it to the
+        // values that might be in the page.  E.g. if we are searching for X IN [5, 3, 7] and five is in pages
+        // 1 and 2 and three is in page 2 and seven is in pages 8 and 9 then when we search page 2 we only need
+        // to search for X IN [5, 3]
```

Comment:
I'm going to leave this as a todo.  I could use the expression simplifier but then I'd have to go from scalar query -> physical expr -> back to scalar query so it would still be a non-tangible amount of code.